grouper/service_account.py

```diff
@@ -199,51 +196,60 @@ def enable_service_account(session, actor, service_account, owner):
 
 
 def service_account_permissions(session, service_account):
-    # type: (Session, ServiceAccount) -> List[ServiceAccountPermission]
-    """Return the permissions of a service account."""
-    permissions = session.query(Permission, ServiceAccountPermissionMap).filter(
+    # type: (Session, ServiceAccount) -> List[ServiceAccountPermissionGrant]
+    """Return the permissions of a service account, including mapping IDs.
+
+    This is used to display the permission grants on a service account page, which has to generate
+    revocation links, so return ServiceAccountPermissionGrant objects that include the mapping ID.
+    """
+    grants = session.query(
+        Permission.name,
+        ServiceAccountPermissionMap.argument,
+        ServiceAccountPermissionMap.granted_on,
+        ServiceAccountPermissionMap.id,
+    ).filter(
         Permission.id == ServiceAccountPermissionMap.permission_id,
         ServiceAccountPermissionMap.service_account_id == service_account.id,
         ServiceAccountPermissionMap.service_account_id == ServiceAccount.id,
         ServiceAccount.user_id == User.id,
         User.enabled == True,
     )
     out = []
-    for permission in permissions:
+    for grant in grants:
         out.append(
-            ServiceAccountPermission(
-                permission=permission[0].name,
-                argument=permission[1].argument,
-                granted_on=permission[1].granted_on,
-                mapping_id=permission[1].id,
+            ServiceAccountPermissionGrant(
+                service_account=service_account,
+                permission=grant.name,
+                argument=grant.argument,
+                granted_on=grant.granted_on,
+                mapping_id=grant.id,
             )
         )
     return out
 
 
 def all_service_account_permissions(session):
-    # type: (Session) -> Dict[str, List[ServiceAccountPermission]]
+    # type: (Session) -> Dict[str, List[PermissionGrant]]
```

Comment:
Hmm, I'm not clear on when `PermissionGrant` is appropriate vs. `ServiceAccountPermissionGrant`. Feels like there ought to be some structuring that avoids awkardness like `service_account_permissions` returning a different type than `all_service_account_permissions`, and like multiple functions wanting to return `PermissionGrant` so filling in `is_alias` with a dummy false

Reply:
Yeah, you're right, this is ugly. This made me finally (after a couple of tries) realize that `PermissionGrant` is indeed superfluous and we can use `GroupPermissionGrant` and `ServiceAccountPermissionGrant` everywhere without too many difficulties. I'll push another patch on this PR that does that.

I think the `is_alias` information is real -- it's just that we have a lot of code that isn't alias-aware yet, so there are no paths to add the aliases. But over time that attribute will become more useful.